docs/reference/aggregations/bucket/categorize-text-aggregation.asciidoc

```diff
@@ -0,0 +1,396 @@
+[[search-aggregations-bucket-categorize-text-aggregation]]
+=== Categorize text aggregation
+++++
+<titleabbrev>Categorize text</titleabbrev>
+++++
+
+experimental::[]
+
+A multi-bucket aggregation that groups semi-structured text into buckets. Each `text` field is re-analyzed
+using a custom analyzer. The resulting tokens are then categorized creating buckets of similarly formatted
+text values. This aggregation works best with machine generated text like system logs.
+
+WARNING: Re-analyzing _large_ result sets will require a lot of time and memory. This aggregation should be
+         used in conjunction with <<async-search, Async search>>.
+
+[[bucket-categorize-text-agg-syntax]]
+==== Parameters
+
+`field`::
+(Required, string)
+The semi-structured text field to categorize.
+
+`max_children`::
+(Optional, integer, default: `100`)
+The maximum number of unique tokens at any given layer of the tokenization tree.
+Must be larger than 1. The smaller the value, the more broad the text categories.
+Larger values may cause the aggregation to more memory and run more slowly
+
+`max_depth`::
+(Optional, integer, default: `5`)
+The maximum number of tokens matched on before attempting to merge categories.
+Larger values may cause the aggregation to more memory and run more slowly.
+
+`similarity_threshold`::
+(Optional, double, default: `0.5`)
+The minimum percentage of tokens that must match for text to be added to the
+category bucket.
+Must be between 0.1 and 1.0. The larger the value the more restrictive the log categories.
+Larger values may increase memory usage.
+
+`categorization_filters`::
```

Comment:
I think it would be good to let people customize the whole analyzer if they want to.  When configuring a categorization job users can choose:

1. Entirely default tokenization and filtering
2. Largely default tokenization and filtering, but with the addition of categorization filters (which are basically pattern char filters that get injected into the default analyzer)
3. Totally custom analyzer

Since you are reusing the `CategorizationAnalyzerConfig` class most of the complexity in having these 3 options will be in code that can be reused.

The parsing bit you'll have to duplicate is in https://github.com/elastic/elasticsearch/blob/15933c0b8ea199de5de3faa8044e8d322138094e/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/job/config/AnalysisConfig.java#L93-L95

And then you need a check that `categorization_filters` and `categorization_analyzer` are not _both_ specified together.

But it doesn't seem _that_ hard to add now.

I was going to say it could be added as a followup but I am wondering if the effort of getting master and 7.x to talk to each other with master having the more complex options and 7.x not will actually be more work than just adding to this PR.  See what you think - if you'd rather add it in a followup PR and you don't think the (temporary) BWC logic will be a nightmare then fair enough.

Reply:
> I think it would be good to let people customize the whole analyzer if they want to. 

Agreed, I can add that.